pandas/core/indexes/period.py

```diff
@@ -665,7 +665,16 @@ def to_timestamp(self, freq=None, how='start'):
         base, mult = _gfc(freq)
         new_data = self.asfreq(freq, how)
 
-        new_data = period.periodarr_to_dt64arr(new_data._values, base)
+        end = how == 'E'
+        if end:
+            indexer = np.where(new_data.notnull())
+            # move forward one period
+            new_data._values[indexer] += 1
+            new_data = period.periodarr_to_dt64arr(new_data._values, base)
+            # subtract one nanosecond
+            new_data[indexer] -= 1
```

Comment:
I think you can just construct the DTI, then ``- pd.Timedelta(1, 'ns')``

note that the display is not correct here

```
In [23]: p = pd.Period('20170101', 'D')

In [24]: p
Out[24]: Period('2017-01-01', 'D')

In [25]: pi = pd.Index([p])

In [26]: pi
Out[26]: PeriodIndex(['2017-01-01'], dtype='period[D]', freq='D')

In [27]: (pi + 1).to_timestamp(how='start') - pd.Timedelta(1, 'ns')
Out[27]: DatetimeIndex(['2017-01-01'], dtype='datetime64[ns]', freq=None)

In [28]: ((pi + 1).to_timestamp(how='start') - pd.Timedelta(1, 'ns')).values
Out[28]: array(['2017-01-01T23:59:59.999999999'], dtype='datetime64[ns]')
```

Reply:
So should the ``DatetimeIndex`` show the time value like below when ``to_timestamp(how='end')`` is called?
```
In [3]: pi.to_timestamp(how='end')
Out[3]: DatetimeIndex(['2017-01-01 23:59:59.999999999'], dtype='datetime64[ns]', freq=None)
```

Reply:
yes, this is a separate bug I think.

Reply:
https://github.com/pandas-dev/pandas/issues/19030